src/java/org/apache/cassandra/index/sai/plan/Plan.java

```diff
@@ -0,0 +1,1504 @@
+    {
+
+        public NumericIndexScan(Factory factory, int id, Expression predicate, long matchingKeysCount)
+        {
+            super(factory, id, predicate, matchingKeysCount);
+        }
+
+        @Override
+        protected double estimateCostPerSkip(double distance)
+        {
+            // This is the first very rough approximation of the cost model for skipTo operation.
+            // It is likely not a very accurate model.
+            // We know for sure that the cost goes up the bigger the skip distance (= the more key we skip over)
+            // and also the bigger the merged posting list is. A range scan of the numeric
+            // index may require merging posting lists from many index nodes. Intuitively, the more keys we match,
+            // the higher number of posting lists are merged. Also, the further we skip, the higher number
+            // of posting lists must be advanced, and we're also more likely to hit a non-cached chunk.
+            // From a few experiments I did, I conclude those costs grow sublinearly.
+            // In the future we probably will need to take more index metrics into account
+            // (e.g. number of distinct values).
+
+            double expectedKeysPerSSTable = expectedKeys() / factory.tableMetrics.sstables;
+            // divide by exponent so the derivative at 1.0 is NUMERIC_INDEX_SKIP_COST_POSTING_LIST_SIZE_FACTOR
+            double dKeys = NUMERIC_INDEX_SKIP_COST_POSTINGS_COUNT_FACTOR / NUMERIC_INDEX_SKIP_COST_POSTINGS_COUNT_EXPONENT;
+            double postingsCountFactor = 1.0 + dKeys * Math.pow(expectedKeysPerSSTable, NUMERIC_INDEX_SKIP_COST_POSTINGS_COUNT_EXPONENT);
+
+            double distanceInPostings = Math.min(expectedKeys(), distance * expectedKeys());
+            // divide by exponent so the derivative at 1.0 is NUMERIC_INDEX_SKIP_COST_DISTANCE_FACTOR
+            double dPostings = NUMERIC_INDEX_SKIP_COST_DISTANCE_FACTOR / NUMERIC_INDEX_SKIP_COST_DISTANCE_EXPONENT;
+            double distanceFactor = 1.0 + dPostings * Math.pow(distanceInPostings, NUMERIC_INDEX_SKIP_COST_DISTANCE_EXPONENT);
+
+            return CostCoefficients.NUMERIC_INDEX_SKIP_COST
+                   * distanceFactor
+                   * postingsCountFactor
+                   * factory.tableMetrics.sstables;
+        }
+    }
+
+    /**
+     * Represents a scan over a literal storage attached index
+     */
+    static class LiteralIndexScan extends IndexScan
+    {
+
+        public LiteralIndexScan(Factory factory, int id, Expression predicate, long matchingKeysCount)
+        {
+            super(factory, id, predicate, matchingKeysCount);
+        }
+
+        @Override
+        protected double estimateCostPerSkip(double distance)
+        {
+            // Literal indexes do not do range scans, so it doesn't matter how many keys we matched -
+            // there is always one posting list. The only thing affecting the cost is how far we skip.
```

Comment:
We can do range queries on maps, so we'll likely want to factor those in at some point. See `MapEntriesIndexTest` for examples.

Reply:
Yes. Also looking at this code now I feel we should likely take the predicate itself into the account. E.g. an equality lookup in a numeric index is probably just as fast as an equality lookup in a literal one. It is range scan what magnifies the number of posting list merges, not the index type. I'll think of taking that into account.  Map scans could be handled differently then as well.

Reply:
Great points. Another relevant detail that we might want to consider in a future iteration is the impact of an index's version on the cost. For example, range queries on maps are particularly inefficient right now due to the on disk format. I have a plan to fix the format, which will trigger a version bump, and then it should be cheaper to run those queries.

Reply:
The map case to be handled in the next milestone when we have good stats.